compiler/rustc_trait_selection/src/solve/search_graph/mod.rs

```diff
@@ -221,27 +232,37 @@ impl<'tcx> SearchGraph<'tcx> {
         canonical_goal: CanonicalGoal<'tcx>,
         mut loop_body: impl FnMut(&mut Self) -> QueryResult<'tcx>,
     ) -> QueryResult<'tcx> {
+        if let Some(result) = tcx.new_solver_evaluation_cache.get(&canonical_goal, tcx) {
+            return result;
+        }
+
         match self.try_push_stack(tcx, canonical_goal) {
             Ok(()) => {}
             // Our goal is already on the stack, eager return.
             Err(response) => return response,
         }
 
-        self.repeat_while_none(
-            |this| {
-                let result = this.deal_with_overflow(tcx, canonical_goal);
-                let stack_elem = this.stack.pop().unwrap();
-                this.try_move_finished_goal_to_global_cache(tcx, stack_elem);
-                result
-            },
-            |this| {
-                let result = loop_body(this);
-                if this.try_finalize_goal(tcx, canonical_goal, result) {
-                    Some(result)
-                } else {
-                    None
-                }
-            },
-        )
+        // This is for global caching, so we properly track query dependencies.
+        // Everything that affects the `Result` should be performed within this
+        // `with_anon_task` closure.
+        //
+        // We only cache the cycle root (why?)
```

Comment:
The `DepGraph` is a DAG, so we cannot give it cycles. Therefore, we need a way to represent this cycle (= a SCC in the evaluation graph) as a tree of `DepNode`s. Those `DepNode`s must contain as transitive dependencies all the queries that are run while evaluating this SCC.

Later, when replaying an execution, the `DepGraph` marks the dependencies green/red in their order of apparition. This is important to avoid ICEs. For instance, if the original execution called `def_kind` to check that a definition was not a closure before calling `fn_sig`, the replay must realise that `def_kind` has changed before trying to force `fn_sig`.

So an SCC's `DepNode`s must contain as transitive dependencies all the queries that are run while evaluating this SCC, *in chronological order*.

Having a single `DepNode` to represent the SCC root (= first goal to have been pushed on the stack) is ok.

Consider a cycle `A -> B -> A` that we cached as a single `DepNode` for the cycle root `A`. The SCC root `DepNode` represents `A`.

When replaying, we know that evaluating `A` makes sense. However, we do not know that the `A -> B` edge nor `B` makes sense from just that information, we need to start re-playing the dependencies to know that.

Conversely, we cannot use the same SCC root `DepNode` to cache `B`. When re-playing, the invoking code would know `B`, and the `DepGraph` would start replaying the execution from `A`. But the edge `B -> A` may not exist any more and `A` may not even make sense.